### ISSUE TITLE
Why we are going with stack-less continuations

### ISSUE THREAD

**Issue by abreslav** [Kotlin/kotlin-coroutines#35]
This should be explained near to the top of the document.

See discussion in #34

**Comment by pron** [comment]
I think you mean "why we *are* going with stack-less continuations" (and not proper delimited continuations)

**Comment by abreslav** [comment]
Fixed, thanks :)

**Comment by abreslav** [comment]
Upsides of full continuations:
- no limitation of only suspending in the immediate body of a coroutine (or an inlined lambda)
- no [red/blue code problem](http://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/)

Downside of known implementations:
- transforming many methods to state machines (may have performance implications?), even if they fiber-block only very rarely;
- bookkeeping through ThreadLocals, which is called every time such a transformed function is invoked;
- if we simply use Quasar, for example, and not add language support, there's no way of turning existing APIs such as Futures into more sequential-looking code.

**Comment by abreslav** [comment]
About optional integration with Quasar: #41

**Comment by pron** [comment]
> transforming many methods to state machines (may have performance implications?), even if they fiber-block only very rarely;

In my experience, transformed methods don't block rarely. The transformed methods are those doing IO or syncrhonization with other threads (say, via channels), and their entire purpose is to block. I am not aware of potentially-blocking functions where the blocking is not a significant part of their main goal.

> bookkeeping through ThreadLocals, which is called every time such a transformed function is invoked;

That is a downside only if the performance impact is significant, but it isn't. These are methods that are meant to block anyway, so the addition of a few nanoseconds to their latency is negligible. OTOH, the use of TLs is an implementation detail which will change on the JVM once native support for continuation is available, and not an issue at all for JavaScript.

> there's no way of turning existing APIs such as Futures into more sequential-looking code.

Why not? [This package](http://docs.paralleluniverse.co/quasar/javadoc/co/paralleluniverse/fibers/futures/package-frame.html) has integrations with `CompletionStage` and Guava's `ListenableFuture` (basically a simple `await` method (we call it `get` but it's just a library function). Language support can be added.

I would say that the downside of existing implementations such as Quasar is that they try to be as general and as transparent as possible, perhaps giving the illusion that any code is fair-game inside a continuation, while there can be some surprises now and again. Hopefully, nearly all of them would go away in Java 9. OTOH, I'm not sure that the existing coroutine spec is free of such surprises. How does it prevent thread-blocking methods from being called from inside a continuation (Quasar detects it at runtime).

**Comment by abreslav** [comment]
> I am not aware of potentially-blocking functions where the blocking is not a significant part of their main goal.

I'd expect most standard library functions, such as `map/flatMap/filter` end up transformed just because they wrap meaningful code too often. And these are not inlined when they operate on sequences (as opposed to collections).

@mikehearn What's your experience?

> How does it prevent thread-blocking methods from being called from inside a continuation (Quasar detects it at runtime).

And what happens when it's detected?

**Comment by mikehearn** [comment]
You get a warning splatted to the console. Hopefully :)

Looking at our (quite small) codebase so far, methods that are @Suspendable but which don't themselves block are just the result of breaking big methods into smaller ones. Standard library methods that run callbacks like map/filter/etc aren't usable with quasar anyway because they aren't annotated, unless they're inlined.

**Comment by pron** [comment]
> I'd expect most standard library functions, such as map/flatMap/filter end up transformed just because they wrap meaningful code too often. 

It is a terrible idea to call blocking code from within such functions. 

> And what happens when it's detected?

It presents a warning (which could be turned off) with a stacktrace of the blocking call. Note that it's hard to do this in general, and hence the careful distinction between continuations and fibers. Quasar currently only detects this with fibers that are scheduled by the default fork-join scheduler. When you have a continuation that does IO and executes on the IO thread, accidental kernel-thread (KT) blocking can have a much more detrimental effect than blocking a fork-join pool's KT, and it is far harder to detect.

**Comment by abreslav** [comment]
> It is a terrible idea to call blocking code from within such functions, anyway.

A prospective style guide entry that looks like a lot of trouble :) 

Could you elaborate on why it's such a bad idea?

**Comment by abreslav** [comment]
> Looking at our (quite small) codebase so far, methods that are @Suspendable but which don't themselves block are just the result of breaking big methods into smaller ones. Standard library methods that run callbacks like map/filter/etc aren't usable with quasar anyway because they aren't annotated, unless they're inlined.

@mikehearn So, all you need is #36, i.e. being able to break a longer coroutine down into smaller chunks in the form of private functions in the same class/file?

**Comment by pron** [comment]
> A prospective style guide entry that looks like a lot of trouble :)

Certainly, but the JDK does have some baggage we have to live with. You gotta take the good with the bad. Note that the solution is the same whether Quasar is used or not: you'd need a separate version of those higher-order functions for blocking use. Language support could help there regardless if Quasar is used or not.

> Could you elaborate on why it's such a bad idea?

Because methods that may block should be designed for blocking. For example, in the JDK, many/most blocking methods throw an `InterruptedException` and/or an `IOException`. In Java, at least, those are checked exceptions that have to be handled in the lambda. Even if they aren't checked, they need to be handled. Also, the JDKs streams API has been specifically designed in such a way that any stream could be executed in parallel. It is debatable whether this was a correct decision or not, but part of the reasoning was precisely that blocking code should be designed in a special way anyway.

The async IO analog to streams is coming to Java 9 in the form of [flows](http://download.java.net/jdk9/docs/api/java/util/concurrent/Flow.html) (based on [reactive streams](https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.0/README.md#specification)), which are then transformed by Quasar (with a streaming analogous to `await`) into channels (or vice-versa).

**Comment by danwallach** [comment]
For what it's worth, it would be exceptionally helpful to keep around stacks, since this lets you do recursion. I tried writing a recursive tree traversal like so:

```
    fun toLazyList(): Sequence<T> = generate<T> {
        if(!empty()) {
            left().toLazyList()
            yield(value())
            right().toLazyList()
        }
    }
```
This didn't work at all. The recursive calls did yield values, but they yielded them into separate sequences which were lost. The following **did** work:
```
    fun toLazyList(): Sequence<T> = generate<T> {
        if(!empty()) {
            left().toLazyList().forEach { yield(it) }
            yield(value())
            right().toLazyList().forEach { yield(it) }
        }
    }
```

(Source code of my attempt to get this to work: https://github.com/danwallach/KLazyTrees)

So, I guess this isn't as bad as I'd feared, but suffice to say that there's a lot of subtlety going on here, and it's hard to look at this program and convince myself that it's appropriately lazy.

**Comment by Groostav** [comment]
Worth mentioning, I was disappointed when I discovered the exact limitations of a stackless implementation (about 20 minutes ago). More comments on the (duplicate) issue I linked to above: #61.

**Comment by elizarov** [comment]
The second revision of coroutines KEEP has full support for stack-full coroutines, albeit current implementation requires an simple adapter to do so (native support of arbitrary code in suspending functions is in the works). I think we can close this issue.

**Comment by elizarov** [comment]
Let me clarify my previous message as the meaning of "coroutines with stack" seems to loosely defined. The second revision of coroutines do not entirely solve red/blue code problem. We still have "regular functions" and "suspending functions" and you can only invoke suspending functions from other suspending functions. The change is that we now _can_ call suspending functions from other suspending functions pretty easy (with a help a simple helper functions as of 1.1-M04) which was much harder in the previous revision (you could do it, but it was syntactically cumbersome). We will make suspending function invocations from other suspending functions completely transparent even in non-tail-call case in the future updates. 

So to summarize, compared to C#/Dart, the second of revision of Kotlin coroutines has less pronounced red/blue problem, because you don't need `await` when calling suspending functions from suspending context. However, compared to Go, Kotlin still clearly suffers from red/blue separation, because in Go _every_ function is effectively suspending. Unfortunately, we cannot go as far as Go (pun intended) due to the need for interop with a regular blocking Java world.
